Improve correctness of CAS_PTR

kdnvt opened the pull request:
This commit have two major changes.

The first change is to set the weak argument in __atomic_compare_exchange to 0, which means that the CAS_PTR can't fail if destination value is equal to expected value.
The implementation of CAS_PTR returns the second variable of __atomic_compare_exchange to its caller, which is expected value if comparison is equal, or destination variable's value if the comparison is unequal. However, the caller can't determine whether the CAS_PTR successes or not if CAS_PTR could fail even if the comparison is equal.

The second change is to set the memory order of CAS_PTR to sequential consistency. The relaxed order will blunder in some situation. For example, if a new node is going to be inserted in linked list:
```
1   new_elem->next = right;
2   if (CAS_PTR(&(left->next), right, new_elem) == right) {
3       FAI_U32(&(the_list->size));
4       return true;
5   }
```
On relaxed order, line 1 may still in write buffer after line 2 finishing, which is invisible to other threads.
There may still have room for improvement on relaxing memory order in
this case, I haven't considered other memory orders carefully.

jserv reviewed (changes requested):
Refine the subject of this pull request and git commit message. The intention was to enforce the sequential consistency order, and you should mention it prior to improving something in advance.

kdnvt commented:
In my opinion, the first change is as important as memory order because it practically influences the correctness of this program, and the bug is harder to spot than memory order. In `include/lock.h`, the function `lock_lock` use below code to implement lock:
```
while (CAS_U32(l, (uint32_t) 0, (uint32_t) 1) == 1)
        ;
```
In case the CAS_PTR fails even if `*l == 0`, the function `lock_lock` could spuriously consider it acquires the lock successfully because CAS_PTR returns 0, but actually, it doesn't acquire the lock.

Should I still refine the subject and commit messages in this case? Or should I split the two changes into different commits if I think these changes are equally important?

jserv commented:
> Should I still refine the subject and commit messages in this case? Or should I split the two changes into different commits if I think these changes are equally important?

Split into two functional commits.

kdnvt committed:
Restrict CAS_PTR to sequential consistency order

Restrict the memory order of CAS_PTR to sequential consistency. The
relaxed order will blunder in some situation. For example, if a new
node is going to be inserted in linked list:

1   new_elem->next = right;
2   if (CAS_PTR(&(left->next), right, new_elem) == right) {
3       FAI_U32(&(the_list->size));
4       return true;
5   }

On relaxed order, line 1 may still in write buffer after line 2
finishing, which is invisible to other threads.
There may still have room for improvement on relaxing memory order in
this case, I haven't considered other memory orders carefully.

kdnvt committed:
Invalidate spurious failure of CAS_PTR

Invalidate the spurious failure of CAS_PTR  by setting the weak
argument of __atomic_compare_exchange to 0.

In original implementation of CAS_PTR, it allows spurious failure,
which means it fails even if the comparison is equal. The caller can
not determine whether the CAS_PTR success or not because it return the
value only by the result of comparison. In include/lock.h , the
function lock_lock use below code to implement lock:

  while (CAS_U32(l, (uint32_t) 0, (uint32_t) 1) == 1)
      ;

In case the CAS_PTR fails even if *l == 0, the function lock_lock could
spuriously consider it acquires the lock successfully because CAS_PTR
returns 0, but actually, it doesn't acquire the lock.
